### ISSUE TITLE
Performance tune Spark

### ISSUE THREAD

**Issue by ehiggs** [hpcugent/hanythingondemand#129]
By default, Spark uses something silly like only 500MB per executor and only three executors. We should add a Spark autogen script to calculate how many executors there should be and how much memory they should get.

**Comment by ehiggs** [comment]
#131 has what appears to be correct parameters for Spark, and in the logs, I see that Spark is asking for the appropriate resources, but for some reason Yarn is not releasing them. There are two issues people are running into:

1. We ask for `spark.executor.instances 2` on a single node, but Yarn only grants one container. If we are on multiple nodes and ask for 5, then we only get 2 (one per node). This limits the cores and memory in use.

2. Some users in a classroom setting have their settings ignored completely and end up with the default Spark memory settings. i.e. 

```
spark.executor.cores 8
spark.executor.instances 2
spark.executor.memory 18432M
``` 

Results in 1 executor  with 512M.

**Comment by ehiggs** [comment]
Executor stdout:

```
2015-12-02 17:06:26,103 INFO  [main] yarn.YarnRMClient (Logging.scala:logInfo(59)) - Registering the ApplicationMaster
2015-12-02 17:06:26,190 INFO  [main] yarn.YarnAllocator (Logging.scala:logInfo(59)) - Will request 2 executor containe
rs, each with 5 cores and 20275 MB memory including 1843 MB overhead
2015-12-02 17:06:26,198 INFO  [main] yarn.YarnAllocator (Logging.scala:logInfo(59)) - Container request (host: Any, ca
pability: <memory:20275, vCores:5>)
2015-12-02 17:06:26,199 INFO  [main] yarn.YarnAllocator (Logging.scala:logInfo(59)) - Container request (host: Any, ca
pability: <memory:20275, vCores:5>)
2015-12-02 17:06:26,232 INFO  [main] yarn.ApplicationMaster (Logging.scala:logInfo(59)) - Started progress reporter th
read with (heartbeat : 3000, initial allocation : 200) intervals
2015-12-02 17:06:27,645 INFO  [Reporter] impl.AMRMClientImpl (AMRMClientImpl.java:populateNMTokens(361)) - Received ne
w token for : node2159.delcatty.data:50680
2015-12-02 17:06:27,650 INFO  [Reporter] yarn.YarnAllocator (Logging.scala:logInfo(59)) - Launching container containe
r_1449072110115_0001_01_000002 for on host node2159.delcatty.data
2015-12-02 17:06:27,652 INFO  [Reporter] yarn.YarnAllocator (Logging.scala:logInfo(59)) - Launching ExecutorRunnable. driverUrl: akka.tcp://sparkDriver@10.141.12.159:42589/user/CoarseGrainedScheduler,  executorHostname: node2159.delcatty.data
2015-12-02 17:06:27,653 INFO  [Reporter] yarn.YarnAllocator (Logging.scala:logInfo(59)) - Received 1 containers from YARN, launching executors on 1 of them.
2015-12-02 17:06:27,653 INFO  [ContainerLauncher #0] yarn.ExecutorRunnable (Logging.scala:logInfo(59)) - Starting Executor Container
```

Yarn ResourceManager:
```
2015-12-02 17:06:21,842 INFO  [ResourceManager Event Processor] scheduler.SchedulerNode (SchedulerNode.java:allocateContainer(141)) - Assigned container container_1449072110115_0001_01_000001 of capacity <memory:2048, vCores:1> on host node2159.delcatty.data:50680, which has 1 containers, <memory:2048, vCores:1> used and <memory:53248, vCores:7> available after allocation
```

So the ResourceManager thinks we have 8 vcores total (1 used; v available) on the master node despite having 16 cores in total there. This 8 comes from the NodeManager:

```
2015-12-02 17:01:50,329 INFO  [main] nodemanager.NodeStatusUpdaterImpl (NodeStatusUpdaterImpl.java:serviceInit(175)) - Initialized nodemanager for null: physical-memory=55296 virtual-memory=116122 virtual-cores=8

...

2015-12-02 17:01:51,541 INFO  [main] nodemanager.NodeStatusUpdaterImpl (NodeStatusUpdaterImpl.java:registerWithRM(304)) - Registered with ResourceManager as node2159.delcatty.data:50680 with total resource of <memory:55296, vCores:8>
```

**Comment by ehiggs** [comment]
This broken default appears to come from Yarn's NodeManager:
```
./hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java:  public static final int DEFAULT_NM_VCORES = 8;
```

**Comment by ehiggs** [comment]
Adding a config override to an `hod clone` version of `IPython-3.2.1` and setting the following fixes the vcores for the 16 core machine I'm working on. 

```
[yarn-site.xml]
yarn.app.mapreduce.am.resource.cpu-vcores=1
yarn.scheduler.maximum-allocation-vcores=16
yarn.scheduler.minimum-allocation-vcoresa=1
yarn.nodemanager.resource.cpu-vcores=16
```

This should be rolled into the autoconfig.

**Comment by ehiggs** [comment]
This has been fixed in #135.
